state/state_test.go

```diff
@@ -1510,6 +1510,27 @@ func (s *StateSuite) TestAllRelations(c *gc.C) {
 	}
 }
 
+func (s *StateSuite) TestAddCloudService(c *gc.C) {
+	svc, err := s.State.AddCloudService(
+		state.AddCloudServiceArgs{
+			Id:         "cloud-svc-ID",
+			ProviderId: "dfsds-sdfsd-sdfds",
+			Addresses:  network.NewAddresses("1.1.1.1"),
+		},
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(svc.Refresh(), jc.ErrorIsNil)
+	c.Assert(state.LocalID(s.State, svc.Id()), gc.Equals, "cloud-svc-ID")
+	c.Assert(svc.ProviderId(), gc.Equals, "dfsds-sdfsd-sdfds")
+	c.Assert(svc.Addresses(), gc.DeepEquals, network.NewAddresses("1.1.1.1"))
+
+	getResult, err := s.State.CloudService("cloud-svc-ID")
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(state.LocalID(s.State, getResult.Id()), gc.Equals, "cloud-svc-ID")
+	c.Assert(getResult.ProviderId(), gc.Equals, "dfsds-sdfsd-sdfds")
+	c.Assert(getResult.Addresses(), gc.DeepEquals, network.NewAddresses("1.1.1.1"))
+}
```

Comment:
Ideally we would add tests that use the transaction Hooks to ensure we are properly doing assertions against the database. (the hooks let you make a change right as the 'transaction' is starting, so that you can see that even though in-memory we think the content is Foo, we properly assert against the database, and notice that the content was changed by someone else to Bar.)

Reply:
very handy, added 2 more tests using hooks for negative cases; thanks